src/operator/tensor/amp_cast.h

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file amp_cast.h
+ * \brief Function definition of casts used by AMP
+ */
+
+#ifndef MXNET_OPERATOR_TENSOR_AMP_CAST_H_
+#define MXNET_OPERATOR_TENSOR_AMP_CAST_H_
+
+#include <vector>
+#include <utility>
+#include <algorithm>
+#include "../mshadow_op.h"
+#include "../mxnet_op.h"
+#include "../elemwise_op_common.h"
+#include "../operator_common.h"
+
+namespace mxnet {
+namespace op {
+
+struct AMPCastParam : public dmlc::Parameter<AMPCastParam> {
+  // use int for enumeration
+  int dtype;
+  DMLC_DECLARE_PARAMETER(AMPCastParam) {
+    DMLC_DECLARE_FIELD(dtype)
+    MXNET_ADD_ALL_TYPES
+    .describe("Output data type.");
+  }
+};
+
+struct AMPMultiCastParam : public dmlc::Parameter<AMPMultiCastParam> {
+  int num_outputs;
+
+  DMLC_DECLARE_PARAMETER(AMPMultiCastParam) {
+    DMLC_DECLARE_FIELD(num_outputs)
+    .describe("Number of input/output pairs to be casted to the widest type.");
+  }
+};
+
+inline bool AMPCastType(const nnvm::NodeAttrs& attrs,
+                        std::vector<int> *in_attrs,
+                        std::vector<int> *out_attrs) {
+  using mshadow::kFloat32;
+  using mshadow::kFloat16;
```

Comment:
Not hard code. It's better to be detected from param.dtype.

Reply:
This is again not really possible to avoid specifying the types here explicitly - you need to be able to figure out whether the type is floating point or not.